### PR TITLE
Deterministic IntegerMath

### PR DESCRIPTION
**Pull request by MaxGraey**
@vgrichina Currently there are pretty basic concept of **IntegerMath** without overflow checking yet. All others functions will be stubs with "not supported" exceptions. WDYT?

Relate to #465

**DONE**
- [x] tests

сс @jannis

### CODE REVIEW

File: std/assembly/math.ts [hunk 1]
```diff
@@ -2287,6 +2287,114 @@ export namespace NativeMathf {
...
+        return <bool>(x >> (sizeof<T>() * 4 - 1));
+      } else {
+        return <T>1;
+      }
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function sqrt<T>(x: T): T {
```
Comment (vgrichina) [inline std/assembly/math.ts]: I think this might benefit from a link to algorithm used.

File: std/assembly/math.ts [hunk 2]
```diff
@@ -2287,6 +2287,114 @@ export namespace NativeMathf {
...
+      if (sizeof<T>() == 8) return <T>ipow64(<i64>x, y);
+      if (sizeof<T>() <= 4) return <T>ipow32(<i32>x, y);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function sign<T>(x: T): T {
+    if (isInteger<T>()) {
+      if (!x) return <T>0;
```
Comment (vgrichina) [inline std/assembly/math.ts]: isn't this short-circuit only needed for unsigned case?

File: std/assembly/math.ts [hunk 3]
```diff
@@ -2287,6 +2287,114 @@ export namespace NativeMathf {
...
+        return 1 - (x >> (sizeof<T>() * 4 - 1) << 1);
+      } else {
+        return <T>1;
+      }
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function signbit<T>(x: T): bool {
```
Comment (vgrichina) [inline std/assembly/math.ts]: Should this delegate into `sign`?

File: std/assembly/math.ts [hunk 4]
```diff
@@ -2287,6 +2287,114 @@ export namespace NativeMathf {
...
+    if (isInteger<T>()) {
+      if (!x) return <T>0;
+      if (isSigned<T>()) {
+        return <bool>(x >> (sizeof<T>() * 4 - 1));
+      } else {
+        return <T>1;
```
Comment (vgrichina) [inline std/assembly/math.ts]: it seems like should be cast to `<bool>` here

File: std/assembly/math.ts (hunk 2 above)
Comment (MaxGraey) [reply std/assembly/math.ts]: Current logic follow this scenario:
1) if `T` is **unsigned** that result will be `1` for all values except `0` which in this case also would `0`.
2) if `T` is **signed** that result will be `-1` for `x < 0`, `1` for `x > 0` and `0` for `0`.

File: std/assembly/math.ts (hunk 3 above)
Comment (MaxGraey) [reply std/assembly/math.ts]: yes, it could be

File: std/assembly/math.ts (hunk 4 above)
Comment (MaxGraey) [reply std/assembly/math.ts]: yash. Good catch! I don't wrote test yet. It's just concept for discussion

File: std/assembly/math.ts (hunk 1 above)
Comment (MaxGraey) [reply std/assembly/math.ts]: I'm not sure about integer `sqrt`. Is it necessary?
Also I could provide  integer `log2` which could be more useful I guess

File: std/assembly/math.ts (hunk 2 above)
Comment (vgrichina) [reply std/assembly/math.ts]: Do you mean that `1 - (x >> (sizeof<T>() * 4 - 1) << 1)` doesn't work for 0?

Comment (MaxGraey) [reply std/assembly/math.ts]: This line of code defines the highest bit and leads to the range [-1, 1] but it can't determine zero from positive value

Comment (vgrichina) [reply std/assembly/math.ts]: nvm then

**Comment by MaxGraey** [comment]
@dcodeIO NaN canonicalization should be done on virtual machine size because this would be faster, safer and don't increase binary size.

**Comment by vgrichina** [comment]
> @dcodeIO NaN canonicalization should be done on virtual machine side because this would be faster, safer and don't increase binary size.

@MaxGraey this is also better for us cause e.g. in distributed consensus context we cannot trust that contract is compiled to canonicalize NaNs, unless it happens either on VM or post-processing stage (vs on compile stage).

**Comment by MaxGraey** [comment]
I will implement this in different way. But current road block is #1137.

Closing for now due to general cleanup described here #1277.

**Pull request closed without merging**